File: JavaScript - p5/Animation.js
class Animation {
    constructor(texture, x, y, w, h, count, rate, offset = 0) {
        this.texture = texture; // the source texture/image/spritesheet
        this.frames = []; // a collection of rectangles that define each frame's location in the texture
        this.rate = rate; // used to control the rate of animation, generally used to slow it down
        this.r = 0.42 * w; // the radius of the sprite; used during collision detection; 42% of the sprite's width
        this.offset = offset; // used to offset the sprite's origin away from its exact center

        // set up the rectangles that define the location of each frame in the texture
        // this works when the sprite's frames are all in a single row
        for (let i = 0; i < count; i++)
            this.frames.push({ x: x + i * w, y: y, w: w, h: h });
    }
}

File: C++ - SFML/Asteroids/src/Animation.cpp
#include "Animation.h"

// a default constructor is necessary in order to declare the animations hash table in Astro.h
Animation::Animation() :
    rate(0), r(0) {
}

Animation::Animation(const sf::Texture& texture, int x, int y, int w, int h, int count, float rate, float offset) :
    rate(rate), r(0.4f * w) {
    sprite.setTexture(texture);
    for (int i = 0; i < count; i++) // set up the rectangles that define the location of each frame in the texture
        frames.push_back(sf::IntRect(x + i * w, y, w, h));

    sprite.setOrigin(static_cast<float>(w * 0.5), static_cast<float>(h * 0.5 + offset)); // move the origin to the center
    sprite.setTextureRect(frames[0]); // set the sprite to show the first frame
}
